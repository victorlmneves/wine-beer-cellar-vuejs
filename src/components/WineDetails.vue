<template>
  <div class="wine">
    <div
      v-if="wineDetails"
      class="wine__item"
    >
      <img class="wine__image"
        :alt="wineDetails.name"
        :src="`https://res.cloudinary.com/vitorneves/image/upload/v1614884127/wineChallenge/${wineDetails.picture}`"
      />
      <div class="wine__info">
        <h3 class="wine__name">{{wineDetails.name}} - {{wineDetails.year}}</h3>
        <h5 class="wine__producer">{{wineDetails.producer}}</h5>
        <p class="wine__description">{{wineDetails.description}}</p>
        <p class="wine__price">{{wineDetails.price | currency}}</p>
      </div>
    </div>
  </div>
</template>

<script>
import { mapActions, mapGetters } from 'vuex'

export default {
  name: 'WineDetails',

  filters: {
    currency: value => {
      if (!value) {
        return
      }

      const options = { style: 'currency', currency: 'EUR' }
      value = value.toString()

      return new Intl.NumberFormat('en-US', options).format(value)
    }
  },

  data () {
    return {
      itemId: null
    }
  },

  computed: {
    ...mapGetters(['wineDetails'])
  },

  mounted() {
    setTimeout(() => {
      window.scrollTo({
        top: 100,
        behavior: 'smooth'
      })
    }, 1500)
  },

  created() {
    this.itemId = this.$route.params.id
    this.getById(this.itemId);
  },

  methods: {
    ...mapActions(['getById'])
  }
}
</script>

<style lang="scss" scoped>
.wine {
  padding: 0 $unit*2 $unit*6 $unit*2;

  &__item {
    margin: 0 $unit $unit*2;
    border: 1px solid #eee;
    padding: $unit*5;
    height: 100%;
    display: flex;
  }

  &__info {
    text-align: left;
    margin-right: $unit*10;
  }

  &__name {
    font-size: $unit*8;
    line-height: 1.2;
    margin-bottom: $unit*3;
  }

  &__producer {
    color: #555;
    font-size: 24px;
    line-height: 1.2;
    margin-bottom: $unit*3;
    margin-top: $unit*3;
    padding-bottom: $unit*6;
    padding-top: $unit*6;
    border-bottom: 1px solid #eee;
    border-top: 1px solid #eee;
  }

  &__price {
    color: #333;
    font-size: 36px;
    line-height: 1;
    margin-top: $unit*6;
  }

  &__image {
    height: 100%;
    max-width: 500px;
    width: 50%;
  }
}
</style>
