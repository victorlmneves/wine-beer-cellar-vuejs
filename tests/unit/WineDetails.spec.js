// import Vue from 'vue'
import Vuex from 'vuex'
import { shallowMount, createLocalVue } from '@vue/test-utils'
import WineDetails from '@/components/WineDetails.vue'
import fakeWines from './fakeData/fakeWines'
// import { actions } from '@/store/modules/wine.js'
// import store from '@/store/index'

const localVue = createLocalVue()
localVue.use(Vuex)

// jest.mock("axios", fakeWines => ({
//   getAll: () => Promise.resolve({ data: fakeWines })
// }))

// class LocalStorageMock {
//   constructor() {
//     this.store = {}
//   }

//   clear() {
//     this.store = {}
//   }

//   getItem(key) {
//     return this.store[key] || null
//   }

//   setItem(key, value) {
//     this.store[key] = String(value)
//   }

//   removeItem(key) {
//     delete this.store[key]
//   }
// }

// global.localStorage = new LocalStorageMock;

// Vue.use(Vuex)

let wrapper = null

let getters = {
  wineDetails: () => fakeWines[2]
}

let actions = {
  getById: jest.fn(() => 2)
}

const $route = {
  params: {
    id: 2
  }
}

const mocks = {
  $route
}

let store = new Vuex.Store({
  getters,
  actions
})

beforeEach(() => {
  wrapper = shallowMount(WineDetails, {
    store,
    localVue,
    mocks
  })
})

afterEach(() => {
  wrapper.destroy()
})

const options = { style: 'currency', currency: 'EUR' }
const unformattedPrice = 33.90
const value = unformattedPrice.toString()
const currency = jest.fn(() => new Intl.NumberFormat('en-US', options).format(value))

describe('WineDetails.vue', () => {
  it('can be mounted correctly', async () => {
    expect(wrapper.element).toMatchSnapshot()
    expect(actions.getById).toHaveBeenCalled()
  })

  it('displays the `N/A` on the price if no value is passed', () => {
    const price = wrapper.find('.wine__price')
    expect(price.text()).toBe('N/A')
  })

  it('displays the item details', () => {
    expect(wrapper.findAll('.wine__item').length).toBe(1)

    const title = wrapper.find('.wine__name')
    expect(title.text()).toBe('Bafarela Reserva - 2018')
  })

  it('displays the price formatted correctly', () => {
    getters = {
      wineDetails: () => fakeWines[1]
    }
    
    actions = {
      getById: jest.fn(() => 1)
    }

    store = new Vuex.Store({
      getters,
      actions
    })

    wrapper = shallowMount(WineDetails, {
      store,
      localVue,
      mocks
    })

    const price = wrapper.find('.wine__price')
    expect(price.text()).toBe('â‚¬27.50')
  })
})
