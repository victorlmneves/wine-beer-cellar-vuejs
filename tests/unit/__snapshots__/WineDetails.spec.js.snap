// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WineDetails.vue can be mounted correctly 1`] = `
<div
  class="wine"
>
  <div
    class="wine__item"
  >
    <img
      alt="Bafarela Reserva"
      class="wine__image"
      src="https://res.cloudinary.com/vitorneves/image/upload/v1614884127/wineChallenge/bafarelaReserva.jpg"
    />
     
    <div
      class="wine__info"
    >
      <h3
        class="wine__name"
      >
        Bafarela Reserva - 2018
      </h3>
       
      <h5
        class="wine__producer"
      >
        Brites Aguiar
      </h5>
       
      <p
        class="wine__description"
      >
        O Bafarela Reserva Tinto 2018 é um vinho de carácter duriense bem marcado que reflete toda a riqueza e qualidade da região. Com uma grande aptidão gastronómica, torna-se numa escolha ideal para antes e durante as refeições.
      </p>
       
      <p
        class="wine__price"
      >
        N/A
      </p>
    </div>
  </div>
</div>
`;
